File: srcs/server/agents/builtin/sleep.go
package builtin

import (
	"context"
	"encoding/json"
	"fmt"
	"time"
)

// SleepTool pauses execution for a specified duration.
// Mirrors CC-Source's SleepTool which allows the agent to yield time
// while waiting for long-running operations (e.g. waiting for CI to finish).
var SleepTool = Tool{
	Name: "Sleep",
	Description: "Sleep for a specified duration. Use this to wait for " +
		"long-running operations (e.g. waiting for CI to complete) rather " +
		"than polling in a tight loop.",
	Parameters: json.RawMessage(`{
		"type": "object",
		"properties": {
			"seconds": {
				"type": "number",
				"description": "Number of seconds to sleep (max 300)."
			}
		},
		"required": ["seconds"]
	}`),
	Execute: func(ctx context.Context, args json.RawMessage) (string, error) {
		var input struct {
			Seconds float64 `json:"seconds"`
		}
		if err := json.Unmarshal(args, &input); err != nil {
			return "", fmt.Errorf("Sleep: invalid args: %w", err)
		}
		if input.Seconds <= 0 {
			return "Slept 0s.", nil
		}
		if input.Seconds > 300 {
			input.Seconds = 300
		}
		dur := time.Duration(input.Seconds * float64(time.Second))
		select {
		case <-time.After(dur):
			return fmt.Sprintf("Slept %.1fs.", input.Seconds), nil
		case <-ctx.Done():
			return "", ctx.Err()
		}
	},
}

File: src/server/src/agents/builtin/tools/sleep.rs
use serde_json::{json, Value};
use std::sync::Arc;

use super::{Tool, ToolExecutor};

// ── Sleep tool ────────────────────────────────────────────────────────────────

struct SleepExecutor;

#[async_trait::async_trait]
impl ToolExecutor for SleepExecutor {
    async fn execute(
        &self,
        args: Value,
    ) -> Result<String, Box<dyn std::error::Error + Send + Sync>> {
        let secs = args["seconds"]
            .as_f64()
            .ok_or("sleep: seconds is required")?;
        let secs = secs.max(0.0).min(60.0); // cap at 60s
        tokio::time::sleep(std::time::Duration::from_secs_f64(secs)).await;
        Ok(format!("Slept for {}s.", secs))
    }
}

pub fn sleep_tool() -> Tool {
    Tool {
        name: "Sleep".to_string(),
        description: "Sleep for a number of seconds (max 60). \
            Use when waiting for async operations."
            .to_string(),
        parameters: json!({
            "type": "object",
            "properties": {
                "seconds": {
                    "type": "number",
                    "description": "Number of seconds to sleep (max 60)."
                }
            },
            "required": ["seconds"]
        }),
        execute: Arc::new(SleepExecutor),
        requires_permission: false,
    }
}
